docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+* Type annotations
+
+In Elisp users are not required to provide type annotations to their
+code.  While at many places the types can be inferred there are
+places, especially in user-defined functions, where we can not infer
+the correct type.
+
+Therefore Elsa provides users with the option to annotate their
+function definitions.  The annotations are placed in the body of the
+function inside a =declare= form:
+
+#+BEGIN_SRC elisp
+(defun add-one (x)
+  (declare (elsa (int) int))
+  (1+ x))
+#+END_SRC
+
+The =declare= form starts with =elsa= followed by one or two values.  If
+two are provided, first must be a list which is the list of input
+types and the second is the return type.
+
+If only one is provided it is taken to be the return type and the
+function must have no input arguments.
+
+Here are general guidelines on how the types are constructed.
+
+** Built-in types
+
+For built-in types with test predicates, drop the =p= or =-p= suffix to
+get the type:
+
+- =stringp= → =string=
+- =integerp= → =integer= (=int= is also accepted)
+- =markerp= → =marker=
+- =hash-table-p= → =hash-table=
+
+Some additional special types:
+
+- =t= stands for =t= and is always true
+- =nil= stands for =nil= and is always false
+- =bool= is a combination of explicitly =t= or =nil=
+
+** Nullable types
+
+By default types do not accept =nil= values.  This helps preventing
+errors where you would pass =nil= to a function expecting a value.  To
+mark a type nullable you can combine it with =nil= type using the [[id:5a21a68a-4df1-4d44-a854-1d9700858a1a][or]]
+combinator: =(or string nil)= is a type which takes any string but also
+=nil=.
+
+** Most general type
+
+A type for everything is called =mixed=.  It accepts anything and is
+always nullable (that means it accepts =nil=).  This is the default type
+for when we lack type information.
+
+** Composite (higher order) types
+
+Composite or higher-order types are types which take other types as
+arguments.
+
+Composite types usually correspond to data constructors such as =cons=,
+=list=, =vector=...
```

Comment:
how do you specify a plist where odd items are symbols and even items are strings? E.g.,:

> `(:background "#583333" :foreground "#29457")

Reply:
For plists I'm planning to have a special type `(plist key value)`, same for alist.

Reply:
Actually it will be a bit more involved since you might also want to express what keys exactly are allowed.  For that I'm pondering a `set` type, such that 

    (set :foo :bar :baz)

Allows any of `:foo`, `:bar`, `:baz` zero or once and nothing else.

Then

    (plist (set :foo :bar :baz) string)

Is a type of a plist with string values and keys `:foo`, `:bar`, `:baz`.

Now this still does not allow us to specify what value type we want for what key type exactly.  In javascript (flow/typescript) they do it simply by enumerating:

```js
{
  name: string,
  age: int
}
```

so maybe we can also have such a syntax:

    (plist :foo string :bar int)


I'm not decided yet here what is the best way.  Ideally somehow combine it all for maximal flexibility.